marlowe-runtime/test/Language/Marlowe/Runtime/Transaction/ConstraintsSpec.hs

```diff
@@ -0,0 +1,734 @@
+  -> [String]
+mustConsumeMarloweOutputViolations MarloweV1 MarloweContext{..} TxConstraints{..} TxBodyContent{..} =
+  case marloweInputConstraints of
+    MarloweInputConstraintsNone -> []
+    MarloweInput invalidBefore invalidHereafter redeemer -> fold
+      [ fromLeft [] do
+          TransactionScriptOutput{..} <- note ["TEST ERROR: MarloweContext doesn't contain script output"] scriptOutput
+          (_, witness) <- note
+            ["Marlowe script UTxO not consumed"]
+            (find ((== utxo) . fromCardanoTxIn . fst) txIns)
+          (witDatum, witRedeemer) <- case witness of
+            BuildTxWith (ScriptWitness _ (PlutusScriptWitness _ _ _ (ScriptDatumForTxIn d) r _)) ->
+              pure (d, r)
+            _ -> Left ["TxInput not build with a plutus script witness"]
+          Left $ fold
+            [ check
+                (witDatum == toCardanoScriptData (toChainDatum MarloweV1 datum))
+                "Input datum does not match"
+            , check
+                (witRedeemer == toCardanoScriptData (Chain.unRedeemer $ toChainRedeemer MarloweV1 redeemer))
+                "Input redeemer does not match"
+            ]
+      , check
+          ( txValidityRange ==
+              ( TxValidityLowerBound ValidityLowerBoundInBabbageEra invalidBefore
+              , TxValidityUpperBound ValidityUpperBoundInBabbageEra invalidHereafter
+              )
+          )
+          "Tx validity range does not match constraints"
+      ]
+
+mustConsumePayoutsViolations
+  :: MarloweVersion v
+  -> MarloweContext v
+  -> TxConstraints v
+  -> TxBodyContent BuildTx BabbageEra
+  -> [String]
+mustConsumePayoutsViolations MarloweV1 MarloweContext{..} TxConstraints{..} TxBodyContent{..} = do
+  roleToken <- Set.toList payoutInputConstraints
+  (("roleToken" <> show roleToken <> ": ") <>) <$> do
+    let
+      isMatch (_, witness) = case witness of
+        BuildTxWith (ScriptWitness _ (PlutusScriptWitness _ _ _ (ScriptDatumForTxIn d) _ _)) ->
+          d == toCardanoScriptData (toChainPayoutDatum MarloweV1 roleToken)
+        _ -> False
+      matchingInputs = fromCardanoTxIn . fst <$> filter isMatch txIns
+      isPayoutUtxo utxo = Map.member utxo payoutOutputs
+    fold
+      [ check (not $ null matchingInputs) "No matching inputs found"
+      , check (all isPayoutUtxo matchingInputs) "Not all matching inputs come from the payout address"
+      ]
+
+requiresSignatureViolations
+  :: MarloweVersion v
+  -> Chain.UTxOs
+  -> TxConstraints v
+  -> TxBodyContent BuildTx BabbageEra
+  -> [String]
+requiresSignatureViolations MarloweV1 utxos TxConstraints{..} TxBodyContent{..} = do
+  pkh <- Set.toList signatureConstraints
+  let
+    inInput ref = case Chain.lookupUTxO ref utxos of
+      Nothing -> False
+      Just Chain.TransactionOutput{..} -> case Chain.paymentCredential address of
+        Just (Chain.PaymentKeyCredential pkh') -> pkh' == pkh
+        _ -> False
+    inExtraKeyWits = case txExtraKeyWits of
```

Comment:
Technically, one doesn't need an extra key witness if the same key witness is already needed to spend one of the UTxOs in the transaction. Including this extra "extra key witness" is harmless in these cases, aside from increasing the transaction size a bit. I don't think we need to change anything here.

Reply:
I'm actually checking `inExtraKeyWits || inInputs` on line 331 below